src/gossip.rs

```diff
@@ -0,0 +1,235 @@
+use crate::error::{MutinyError, MutinyStorageError};
+use crate::logging::MutinyLogger;
+use anyhow::anyhow;
+use bitcoin::Network;
+use bitcoin_hashes::hex::{FromHex, ToHex};
+use lightning::util::ser::{ReadableArgs, Writeable};
+use log::{debug, info, warn};
+use reqwest::Client;
+use rexie::{ObjectStore, Rexie, TransactionMode};
+use std::sync::Arc;
+use wasm_bindgen::JsValue;
+
+use crate::node::{NetworkGraph, RapidGossipSync};
+use crate::wallet::get_rgs_url;
+
+pub(crate) const GOSSIP_DATABASE_NAME: &str = "gossip";
+pub(crate) const GOSSIP_OBJECT_STORE_NAME: &str = "gossip_store";
+pub(crate) const GOSSIP_SYNC_TIME_KEY: &str = "last_sync_timestamp";
+pub(crate) const NETWORK_GRAPH_KEY: &str = "network_graph";
+
+struct Gossip {
+    pub last_sync_timestamp: u32,
+    pub network_graph: NetworkGraph,
+}
+
+impl Gossip {
+    pub fn new(network: Network, logger: Arc<MutinyLogger>) -> Self {
+        Self {
+            last_sync_timestamp: 0,
+            network_graph: NetworkGraph::new(network, logger),
+        }
+    }
+}
+
+async fn build_gossip_database() -> Result<Rexie, MutinyError> {
+    // Create a new database
+    let rexie = Rexie::builder(GOSSIP_DATABASE_NAME)
+        // Set the version of the database to 1.0
+        .version(1)
+        .add_object_store(ObjectStore::new(GOSSIP_OBJECT_STORE_NAME))
+        // Build the database
+        .build()
+        .await?;
+
+    Ok(rexie)
+}
+
+async fn get_gossip_data(rexie: &Rexie, logger: Arc<MutinyLogger>) -> Result<Gossip, MutinyError> {
+    // Create a new read-only transaction
+    let transaction = rexie.transaction(&[GOSSIP_OBJECT_STORE_NAME], TransactionMode::ReadOnly)?;
+
+    let store = transaction.store(GOSSIP_OBJECT_STORE_NAME)?;
+
+    // Get the `last_sync_timestamp`
+    let last_sync_timestamp_js = store.get(&JsValue::from(GOSSIP_SYNC_TIME_KEY)).await?;
+    let last_sync_timestamp: u32 =
+        last_sync_timestamp_js
+            .as_f64()
+            .ok_or_else(|| MutinyError::ReadError {
+                source: MutinyStorageError::Other(anyhow!(
+                    "could not read last_sync_timestamp, got {:?}",
+                    last_sync_timestamp_js
+                )),
+            })? as u32;
```

Comment:
Should we just default to 0 here to not cause any client errors? Worst case it fetches it down entirely but that's just 4mb or so, not the end of the world.

Reply:
On line 123 we do this, if this function errors we just use a fresh Network Graph and download the whole thing